File: prova2/codificacao.hs
-- Função de codificar uma String
codificar :: String -> String
codificar [] = []
codificar (x : xs) = codificarAux (getStringOfEqualsChars (x : xs) ' ')

-- Codifica uma lista de Strings de caracteres iguais em uma String (Função auxiliar)
codificarAux :: [String] -> [Char]
codificarAux [] = []
codificarAux (x : xs) = int2str (countEqualsChars x) ++ firstItemOfList x ++ codificarAux xs

-- Retorna a lista separada de Strings de caracteres iguais
getStringOfEqualsChars :: [Char] -> Char -> [String]
getStringOfEqualsChars [] ' ' = []
getStringOfEqualsChars (x : xs) returnItem
  | isLastItem = [] -- Verifica se é o último item da lista
  | returnItem == x && checkListaVazia = getStringOfEqualsChars xs x -- Verifica se o item atual é igual ao item anterior e se a lista não está vazia
  | [x] == firstItemOfList xs = linearizar [[x : auxGetStringOfEqualsChars x xs]] ++ getStringOfEqualsChars xs x 
  | otherwise = [x] : getStringOfEqualsChars xs ' '
  where
    isLastItem
      | length xs == 0 && x == returnItem = True -- Verifica se é o último item da lista
      | otherwise = False
    checkListaVazia
      | xs == [] = False -- Verifica se a lista está vazia
      | otherwise = True

-- Retorna a lista de Strings de caracteres iguais agrupando os caracteres iguais
auxGetStringOfEqualsChars :: Char -> [Char] -> [Char]
auxGetStringOfEqualsChars c [] = []
auxGetStringOfEqualsChars c (x : xs)
  | c == x = x : auxGetStringOfEqualsChars c xs
  | otherwise = []

-- Copia o primeiro item de uma Lista e adiciona na próxima
linearizar :: [[String]] -> [String]
linearizar [] = []
linearizar (x : xs) = x ++ linearizar xs

-- Retorna o primeiro item de uma lista
firstItemOfList :: String -> String
firstItemOfList [] = ""
firstItemOfList (x : xs) = [x]

-- Retorna o primeiro char de uma lista
firstCharOfList :: String -> Char
firstCharOfList [] = ' '
firstCharOfList (x : xs) = x

-- Retorna a quantidade de caracteres iguais em uma String que está em uma lista
countEqualsChars :: String -> Int
countEqualsChars [] = 0
countEqualsChars (x : xs) = length (x : xs)

-------------------------------------------------------------------

-- Função de decodificar uma String
decodificar :: String -> String
decodificar [] = []
decodificar (x : xs)
  | isDigit x && nextDigitIsNotDigit = decodificarAux (str2int digit) (firstItemOfList xs) ++ decodificar xs -- Caso só um dígito
  | isDigit x && nextDigitIsDigit = decodificarAux (str2int digit) (firstItemOfList (nextItemOfList xs)) ++ decodificar (nextItemOfList xs) -- Caso mais de um dígito
  | otherwise = decodificar xs
  where
    nextDigitIsDigit = isDigit (firstCharOfList xs) -- Verifica se o próximo char depois do primeiro dígito é um número
    nextDigitIsNotDigit | nextDigitIsDigit = False | otherwise = True -- Verifica se o próximo char depois do primeiro dígito não é um número
    digit
      | nextDigitIsDigit = [x] ++ firstItemOfList xs -- Se o próximo char depois do primeiro dígito for um número, concatena o primeiro dígito com o próximo char
      | otherwise = [x] -- Se o próximo char depois do primeiro dígito não for um número, retorna apenas o primeiro dígito

-- Retorna o próximo item de uma lista
nextItemOfList :: String -> String
nextItemOfList [] = []
nextItemOfList (x : xs) = xs

-- Função auxiliar de decodificar uma String
decodificarAux :: Int -> String -> String
decodificarAux 0 cs = []
decodificarAux i c = c ++ decodificarAux (i - 1) c

-------------------------------------------------------------------

-- Funções auxiliares dadas no enunciado
isDigit :: Char -> Bool
isDigit c = c >= '0' && c <= '9'

int2str :: Int -> String
int2str i = show i

str2int :: String -> Int
str2int s = read s

-------------------------------------------------------------------

File: prova2/codificacao.pl
is_digit(X) :- atom_length(X,1), X @>= '0', X @=< '9'.

concatenar([], L, L).
concatenar([X|L1], L2, [X|L3]) :- concatenar(L1, L2, L3).
  
linearizar([], []).
linearizar([X|L], LR) :- linearizar(L, LT), concatenar(X, LT, LR).
 
compactar([], []).
compactar([X], [[1,X]]) :- !.
compactar([X,X|L1], [[N,X]|L2]) :- compactar([X|L1], [[Z,X]|L2]), N is Z+1.
compactar([X,Y|L1], [[1,X]|L2]) :- not(X == Y), compactar([Y|L1], L2).

codificarAux([], []).
codificarAux([X,Y|L], R) :-  codificarAux(L, T1), number_chars(X, T2), concatenar(T2, [Y], T3), concatenar(T3, T1, R).

codificar(S, C) :- string_chars(S, T1),
                   compactar(T1, T2),
                   linearizar(T2, T3),
                   codificarAux(T3, T4),
                   string_chars(C, T4).
repetir(0, _, []).
repetir(N, C, R) :- N1 is N - 1, repetir(N1, C, T1), concatenar(T1, [C], R).

decodificarAux(_, [], []).
decodificarAux(DS, [X|L], R) :- is_digit(X), concatenar(DS, [X], T1), decodificarAux(T1, L, R).
decodificarAux(DS, [X|L], R) :- not(is_digit(X)), number_chars(N, DS), repetir(N, X, T1), decodificarAux([], L, T2), concatenar(T1, T2, R).

decodificar(S, D) :- string_chars(S, T1), decodificarAux([], T1, T2), string_chars(D, T2).